GTO: updating docs (#235)
iterative/mlem.ai: content/docs/gto/user-guide/dvc.md

# GTO with DVC

> In many ML projects, data isn't stored in a Git repository and needs to be
> downloaded from external sources. [DVC](https://dvc.org) is a common way to
> store binaries for the artifacts registered with GTO.

Reviewer: ```suggestion
Large files are typically not stored in a Git repository, so they need to be
downloaded from external sources. [DVC](https://dvc.org) is a great way to
store your GTO artifact files while keeping a pointer in the
repo, and simplifying [data management] and synchronization.

[data management]: https://dvc.org/doc/user-guide/data-management
```

<details>

### Learn about different approaches to this

1. You can commit artifacts to Git repo. If they aren't small enough, this is
   not recommended. To bypass this limitation, you can use
   [Git-lfs](https://git-lfs.github.com).
2. You can version binaries with [DVC](https://dvc.org/) and commit pointers to
   them to the repo. This is the recommended approach for large files.
3. You can version binaries manually somewhere, specifying URL to them as `path`
   in `artifacts.yaml`. This can be done, if your artifacts are already

>    versioned by some external system.

Reviewer: Many Qs:

- Why is this hidden?
- `You can commit artifacts to Git repo` - is that a realistic scenario? Let's not even mention it otherwise.
- Mention DVC before Git LFS.
- Why are we talking about data versioning? Seems to me it complicates the explanation.
- Are artifacts always "binaries" ? Maybe just call them "artifact files", "large files", "data", or something like that (more generic).
- `versioned by some external system` - what are we referring to here? (Examples)

Reviewer: Kk 🙂

</details>

> If you are new to DVC, you should start with
> [DVC Get Started](https://dvc.org/doc/start) first, and then come back to this
> Guide.
> 
> <!-- ```
> 
> dvc init --no-scm dvc remote add az azure://container-name export
> AZURE_STORAGE_CONNECTION_STRING='YOUR_CONNECTION_STRING'
> 
> dvc import-url --no-download azure://container-name/data.parquet
> 
> ``` -->

Reviewer: Admon, simplified text, remove HTML comment:

```suggestion
<admon icon="book">

If you're new to DVC,
[get started here](https://dvc.org/doc/start) first.

</admon>
```

## Tracking an artifact with DVC

> First, we need to start tracking artifact with DVC. If you produce this artifact

Reviewer: I don't think we need this section at all: it's DVC documentation (redundant since we already linked to the Get Started)...

Author: Besides, this shows `dvc import-url`, which could be an important scenario for some folks starting to use MR while having some models outside.

Author: People without context will get the bare minimum from this part. People with context will just see some story here.

Author: WDYT?

Reviewer: I think that linking to https://dvc.org/doc/start/data-management/data-versioning or even straight to https://dvc.org/doc/command-reference/add would be acceptable.

Not a super strong opinion since this section is small, but I think that in principle we should not cross-document products.

_Logic: what if the DVC feature changes? dvc.org/doc will probably get updated, but not here._

> Besides, this shows dvc import-url

Not seeing why that's particularly important. BTW, the blocks are missing `git add .gitignore`

in DVC Pipelines, it's done automatically.

If the artifact is located inside your Git repo, you can use `dvc add`:

```cli
$ dvc add model.pkl
$ git add model.pkl.dvc
```

If the artifact is located in some external storage, we can use `dvc import-url`
to still keep metainformation about it in the repo (use `--no-download` to skip
downloading it):

```cli
$ dvc import-url --no-download s3://container/model.pkl
$ git add model.pkl.dvc
```

## Annotating DVC-tracked artifacts

> Once the artifact is tracked with DVC within your repo, we can annotate it with
> GTO:

Reviewer: ... Instead, just link to DVC docs for specific operations of relevance, like this:

```suggestion
Once the artifact is
[tracked with DVC](https://dvc.org/doc/command-reference/add) within your repo,
`gto annotate` it:
```

p.s. remember to try and use the commands in-line at least once (not just in fenced code blocks) so they get auto-linked (to the cmd ref).

```cli
$ gto annotate model --path model.pkl
```

> This will modify `artifacts.yaml`, adding:

Reviewer: Copy edit
```suggestion
This will write the following to `artifacts.yaml`:
```

Reviewer: Also, Maybe mention `model.pkl.dvc` somewhere?

```yaml
model:
  path: model.pkl
```

> Now you should commit changes to Git, and you can register versions and assign
> stages referencing the new commit.
> 
> ```cli
> $ git add artifacts.yaml
> $ git commit -m "version artifact binaries with DVC and annotate it with GTO"

Reviewer: ````suggestion
Commit the changes to Git in order to `gto register` artifact versions and
`gto assign` them to deployment stages referencing the new commit.

```cli
$ git add artifacts.yaml
$ git commit -m "version artifact binaries with DVC and annotate it with GTO"
````

Reviewer: Why do you want to reference a commit in `gto register/assign` though? I'm not sure this is clear (maybe link to the UG index page section where this is explained? Or elaborate here perhaps)

> $ git push
> ```
> 
> Now your changes is live in your Git repo and you can download your artifact to
> use it 🙌

Reviewer: I feel like the pushing part should be separate. It's kind of a different step (compared to `git commit` + possibly `gto register/assign`. Also, we're missing `dvc push`? Idea:

````suggestion
```

To share your work, you'll need [remote storage] setup in DVC. You can then
upload the artifact files and the changes to the repo:

[remote storage]: https://dvc.org/doc/command-reference/remote

```cli
$ dvc push
$ git push
```
````

Note that we're assuming that there's a Git remote. Is this safe?

Author: yes, this doesn't need to be 100% reproducible I think

## Downloading artifacts

> When you want to download GTO artifact which binaries are stored with DVC, you
> need to use `dvc get` or `dvc import` command:

Reviewer: Copy edit
```suggestion
To download GTO artifact files tracked with DVC, you
can use the `dvc get` or `dvc import` commands.
```
But...

Reviewer: Wait, what about `dvc pull`? Seems like we're missing that scenario (which is the simplest). Maybe cover that and only mention+link to the `dvc import/get` refs for more advanced uses.

Author: To use `dvc pull`, you need to be inside of the repo. In many cases, you just will want to get the artifact without cloning the repo. But I've added `dvc pull` mention, thanks for the suggestion.

Reviewer: Thanks

```cli
$ dvc get $REPO $ARTIFACT_PATH --rev $REVISION -o $OUTPUT_PATH
```

Check out [User Guide](/doc/gto/user-guide#getting-artifacts-downstream) to
learn how to find out `ARTIFACT_PATH` and `REVISION`.

<details>

### Example: downloading from outside the repo

If you need to download the latest version of `model`, that would be:

```cli
$ REVISION=$(gto show --repo $REPO model@greatest --ref)
$ ARTIFACT_PATH=$(gto describe --repo $REPO $ARTIFACT --rev $REVISION --path)
$ dvc get $REPO $ARTIFACT --rev $REVISION -o $ARTIFACT_PATH
```

</details>

<details>

### Example: downloading in CI

If you need to download an artifact from the same repo, that would be a bit
simpler (taking GH Actions as an example):

```cli
$ ARTIFACT_PATH=$(gto describe model --rev $GITHUB_REF --path)

> $ dvc pull $ARTIFACT_PATH

Reviewer: > from the same repo

Well, it can be a distributed copy of the repo that anyone else has! This is the happy path no? Straightforward Git collaboration/ GitOps.

Author: Not sure what to add to reflect that... Can you maybe suggest something?

Reviewer: This CI example is good. I just meant that the wording may be confusing. Is it "the same repo" or is it another copy of the repo (in the CI machine)? I guess it's clear from the context, nvmd.

```

</details>

> 
> <!--
> Refer to DVC install guide and Get Started to learn DVC first. If you're already
> familiar with DVC, let's set it up for GTO repo (use your remote storage instead
> of `s3://mybucket/myrepo`):
> 
> ```cli
> $ dvc init
> $ dvc remote add myremote -d s3://mybucket/myrepo
> $ git add .dvc/config
> ```
> 
> ##
> 
> If you want to version your artifact with DVC, you need to add it to DVC first
> (skip this step if you already have DVC-tracked artifacts):
> 
> ```cli
> $ dvc add path/to/artifact
> $ dvc push
> $ git add
> ``` -->

Reviewer: No need, all this is DVC docs. Doesn't belong here.
```suggestion
```
